### PR TITLE
feat: Database Seeding Functionality

### PR DESCRIPTION
**Pull request by kkumar-gcc**
Issue: https://github.com/goravel/goravel/issues/66

## 📑 Description
It implements the seeding functionality by providing users with the freedom to choose which seeders to execute. It introduces a configuration mechanism where users can specify the seeders they want to run, allowing for more flexibility and customization.

## ✅ Checks

- [x] My code requires changes to the documentation
- [x] My pull request adheres to the code style of this project
- [ ] I have updated the documentation as required
- [ ] All the tests have passed

### CODE REVIEW

**Comment by hwbrzzl** [comment]
Could you please update some use cases in the description?

File: database/console/factory_make_command.go
Comment (hwbrzzl) [inline database/console/factory_make_command.go:1]: S: I suggest we implement the factory feature in another PR after this one.

File: database/seeder.go [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+//	seeder.Call([]interface{}{&UserSeeder{}, &PostSeeder{}}, true, nil)
+//
+// Parameters:
+//   - class (interface{}): The seeder class or a slice of seeder classes to execute.
+//   - silent (bool): Determines whether the execution should be silent or display output.
+//   - parameters ([]interface{}): Optional. Additional parameters to pass to the seeder(s).
+//
+// Returns:
+//   - error: An error if the execution fails.
+func (s *Seeder) Call(class interface{}, silent bool, parameters []interface{}) error {
```
Comment (hwbrzzl) [inline database/seeder.go:59]: 1. The `parameters` can be passed directly if needed: `seeder.Call([]seeder.Seeder{&UserSeeder{Parameters: *}})`;
2. I think `silent` can not be supported now;
```suggestion
func (s *Seeder) Call(seeders []seeder.Seeder) error {
```

**Review by hwbrzzl — COMMENTED** [review]
A good start 👍

File: contracts/database/seeder.go
[the remark this answers is not in the document]
Comment (kkumar-gcc) [reply contracts/database/seeder.go:1]: done

File: contracts/database/seeder.go [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+	// Register registers seeders.
+	Register(seeders []Seeder)
+
+	// Run executes the seeder logic.
+	Run(ctx console.Context) error
+
+	// SetCommand sets the console command instance on the seeder.
+	SetCommand(command console.Context)
+
+	GetSeeder(name string) Seeder
```
[the remark this answers is not in the document]
Comment (kkumar-gcc) [reply contracts/database/seeder.go:17]: Okay, i have removed SetCommand but GetSeeder is needed for exposing an instance of seeder so i have added it in seeder Facade.

File: database/seeder.go (hunk 1 above)
Comment (kkumar-gcc) [reply database/seeder.go:59]: Done

File: database/console/factory_make_command.go
Comment (kkumar-gcc) [reply database/console/factory_make_command.go:1]: Okay

File: database/console/seed_command.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+}
+
+// Extend The console command extend.
+func (receiver *SeedCommand) Extend() command.Extend {
+	return command.Extend{
+		Category: "db",
+		Flags: []command.Flag{
+			&command.StringFlag{
+				Name:    "seeder",
+				Value:   "DatabaseSeeder",
```
Comment (hwbrzzl) [inline database/console/seed_command.go:47]:
```suggestion
				Value:   "",
```

File: database/console/seed_command.go [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+		log.Println("No valid seeder instance found.")
+		return nil
+	}
+	seeder.Run()
+	return nil
+}
+
+// ConfirmToProceed determines if the command should proceed based on user confirmation.
+func (receiver *SeedCommand) ConfirmToProceed(ctx console.Context) bool {
+	force := ctx.Option("force")
```
Comment (hwbrzzl) [inline database/console/seed_command.go:74]: S: You didn't add the force option and you can use `OptionBool` to get the value.

File: database/console/seed_command.go [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+	alert := receiver.config.GetString("alert")
+	cancel := receiver.config.GetString("cancel")
+
+	fmt.Println(alert)
+
+	confirmed := receiver.config.GetBool("confirmed")
+	if !confirmed {
+		fmt.Println(cancel)
+		return false
+	}
```
Comment (hwbrzzl) [inline database/console/seed_command.go:93]: S: We don't have this function, we can run the command directly without confirming in PROD if force is true.

File: database/console/seed_command.go [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+	return true
+}
+
+// GetSeeder returns a seeder instance from the container.
+func (receiver *SeedCommand) GetSeeder(ctx console.Context) seeder.Seeder {
+	class := ctx.Argument(0)
+	if class == "" {
+		class = ctx.Option("seeder")
+	}
+	class = "seeders." + class
```
Comment (hwbrzzl) [inline database/console/seed_command.go:103]: S: `class` can be empty, run all seeders when `class` is empty.

File: database/console/seed_command.go [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+	instance, err := receiver.app.Make("goravel.seeder")
+	if err != nil {
+		log.Println("Failed to resolve seeder instance:", err)
+		return nil
+	}
+	seeders, ok := instance.(seeder.Facade)
```
Comment (hwbrzzl) [inline database/console/seed_command.go:109]: S: Inject `facades.Seeder` like `config.Config`.

File: database/console/seeder_make_command.go [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
+func (receiver *SeederMakeCommand) getStub(name string) string {
+	return Stubs{}.Seeder(name)
+}
+
+// getPath Get the full path to the command.
+func (receiver *SeederMakeCommand) getPath(name string) string {
+	pwd, _ := os.Getwd()
+
+	return pwd + "/database/seeders/" + str.Camel2Case(name) + ".go"
```
Comment (hwbrzzl) [inline database/console/seeder_make_command.go:64]: S: Need to support subdirectory.

File: database/console/stubs.go [hunk 9]
```diff
@@ -71,3 +75,54 @@ func (u *DummyObserver) ForceDeleted(event orm.Event) error {
...
 }
+
+func (r Stubs) Factory(name string) string {
+	return fmt.Sprintf(`package factories
+
+import (
+	"github.com/goravel/framework/database/orm"
+)
+
+type %s struct {
```
Comment (hwbrzzl) [inline database/console/stubs.go:86]: S: Please use `DummyPackage` like other stubs, the same as below.

File: database/seeder.go [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+//   - seeders ([]seeder.Seeder): The seeder class or a slice of seeder classes to execute.
+//
+// Returns:
+//   - error: An error if the execution fails.
+func (s *Seeder) Call(seeders []seeder.Seeder) error {
+	for _, seeder := range seeders {
+		name := fmt.Sprintf("%T", seeder)
+
+		if contains(s.Called, name) {
+			continue
```
Comment (hwbrzzl) [inline database/seeder.go:56]: Q: Why pass the seeder when contains?

File: database/seeder.go [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+//	seeder := &Seeder{}
+//	seeder.Call([]seeder.Seeder{&UserSeeder{}})
+//	seeder.Call([]seeder.Seeder{&UserSeeder{}, &PostSeeder{}})
+//
+// Parameters:
+//   - seeders ([]seeder.Seeder): The seeder class or a slice of seeder classes to execute.
+//
+// Returns:
+//   - error: An error if the execution fails.
+func (s *Seeder) Call(seeders []seeder.Seeder) error {
```
Comment (hwbrzzl) [inline database/seeder.go:51]:
```suggestion
func (s *Facade) Call(seeders []seeder.Seeder) error {
```

File: database/seeder.go [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+	"fmt"
+	"log"
+	"reflect"
+
+	"github.com/goravel/framework/contracts/database/seeder"
+)
+
+type Seeder struct {
+	Called []string
+}
```
Comment (hwbrzzl) [inline database/seeder.go:13]: S: It seems we don't need this. We can use `facades.Seeder().Call()` directly in a seeder.

File: database/console/seed_command.go (hunk 3 above)
Comment (kkumar-gcc) [reply database/console/seed_command.go:47]: done

File: database/console/seed_command.go (hunk 4 above)
Comment (kkumar-gcc) [reply database/console/seed_command.go:74]: done

File: database/console/seed_command.go (hunk 5 above)
Comment (kkumar-gcc) [reply database/console/seed_command.go:93]: done

File: database/seeder.go (hunk 12 above)
Comment (kkumar-gcc) [reply database/seeder.go:13]: We are using `Called` array to keep track of called seeder for CallOnce function

File: database/console/seed_command.go (hunk 7 above)
Comment (kkumar-gcc) [reply database/console/seed_command.go]: done

File: database/console/seed_command.go [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+	"github.com/goravel/framework/contracts/config"
+	"github.com/goravel/framework/contracts/console"
+	"github.com/goravel/framework/contracts/console/command"
+	"github.com/goravel/framework/contracts/database/seeder"
+	"github.com/goravel/framework/contracts/foundation"
+)
+
+type SeedCommand struct {
+	config config.Config
+	app    foundation.Application
```
Comment (hwbrzzl) [inline database/console/seed_command.go]: S: We should inject `seeder seeder.Facade` instead of `app`.

File: database/console/seed_command.go [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+func (receiver *SeedCommand) ConfirmToProceed(ctx console.Context) bool {
+	force := ctx.OptionBool("force")
+	if force || (receiver.config.Env("APP_ENV") != "production") {
+		return true
+	}
+
+	// Display production alert message
+	receiver.config.Add("alert", "Application In Production")
+
+	alert := receiver.config.GetString("alert")
```
Comment (hwbrzzl) [inline database/console/seed_command.go:84]: Q: Why add `alert` to the config and get it from the config?

File: database/console/seed_command.go [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+
+	color.Yellowln(alert)
```
Comment (hwbrzzl) [inline database/console/seed_command.go:86]: S: it's better to return an error and print the error in the `Handle` method.

File: database/console/seed_command.go [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+func (receiver *SeedCommand) Extend() command.Extend {
+	return command.Extend{
+		Category: "db",
+		Flags: []command.Flag{
+			&command.BoolFlag{
+				Name:    "force",
+				Aliases: []string{"f"},
+				Usage:   "force the operation to run when in production",
+			},
+			&command.StringFlag{
```
Comment (hwbrzzl) [inline database/console/seed_command.go:48]: S: This can be `StringSliceFlag`.

File: database/console/seed_command.go [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+}
+
+// Handle executes the console command.
+func (receiver *SeedCommand) Handle(ctx console.Context) error {
+	if !receiver.ConfirmToProceed(ctx) {
+		return nil
+	}
+
+	color.Greenln("Seeding database.")
+	seeder := receiver.GetSeeder(ctx)
```
Comment (hwbrzzl) [inline database/console/seed_command.go:65]: S: Should pass `[]string` to `GetSeeder`.

File: database/service_provider.go [hunk 18]
```diff
@@ -34,6 +38,7 @@ func (database *ServiceProvider) Boot(app foundation.Application) {
 
 func (database *ServiceProvider) registerCommands(app foundation.Application) {
 	config := app.MakeConfig()
+	facade := app.MakeSeeder()
```
Comment (hwbrzzl) [inline database/service_provider.go:41]:
```suggestion
	seeder := app.MakeSeeder()
```

File: database/console/seed_command.go [hunk 19]
```diff
@@ -0,0 +1,109 @@
...
+	force := ctx.OptionBool("force")
+	if force || (receiver.config.Env("APP_ENV") != "production") {
+		return nil
+	}
+	return errors.New("application in production use --force to run this command")
+}
+
+// GetSeeder returns a seeder instance from the container.
+func (receiver *SeedCommand) GetSeeders(names []string) []seeder.Seeder {
+	seeders := receiver.seeder
```
Comment (hwbrzzl) [inline database/console/seed_command.go:93]: S: We can use `receiver.seeder` directly.

File: contracts/database/seeder/seeder.go [hunk 20]
```diff
@@ -0,0 +1,16 @@
...
+
+type Facade interface {
+	// Register registers seeders.
+	Register(seeders []Seeder)
+
+	// GetSeeder gets a seeder instance from the seeders.
+	GetSeeder(name string) Seeder
+
+	// All seeders
+	GetAllSeeder() []Seeder
```
Comment (hwbrzzl) [inline contracts/database/seeder/seeder.go:11]:
```suggestion
	GetSeeders() []Seeder
```

File: database/console/seeder_make_command_test.go [hunk 21]
```diff
@@ -0,0 +1,23 @@
...
+	seederMakeCommand := &SeederMakeCommand{}
+	mockContext := &consolemocks.Context{}
+	mockContext.On("Argument", 0).Return("").Once()
+	assert.Nil(t, seederMakeCommand.Handle(mockContext))
+	assert.False(t, file.Exists("database/seeders/user_seeder.go"))
+
+	mockContext.On("Argument", 0).Return("UserSeeder").Once()
+	assert.Nil(t, seederMakeCommand.Handle(mockContext))
+	assert.True(t, file.Exists("database/seeders/user_seeder.go"))
+	assert.True(t, file.Remove("database"))
```
Comment (hwbrzzl) [inline database/console/seeder_make_command_test.go:22]: S: Add test cases for subdirectory.

File: database/console/stubs.go [hunk 22]
```diff
@@ -71,3 +71,54 @@ func (u *DummyObserver) ForceDeleted(event orm.Event) error {
 }
 `
 }
+
+func (r Stubs) Factory(name string) string {
```
Comment (hwbrzzl) [inline database/console/stubs.go:75]: S: Remove this function first.

File: database/console/stubs.go [hunk 23]
```diff
@@ -71,3 +71,54 @@ func (u *DummyObserver) ForceDeleted(event orm.Event) error {
...
+
+func (r Stubs) Seeder() string {
+	return `package DummyPackage
+
+import (
+	"github.com/goravel/framework/database"
+)
+	
+type DummySeeder struct {
+	database.Seeder
```
Comment (hwbrzzl) [inline database/console/stubs.go:96]:
```suggestion
```

File: database/seeder.go [hunk 24]
```diff
@@ -0,0 +1,109 @@
...
+
+type Facade struct {
```
Comment (hwbrzzl) [inline database/seeder.go:15]:
```suggestion
type SeederFacade struct {
```

File: contracts/database/seeder/seeder.go [hunk 25]
```diff
@@ -0,0 +1,16 @@
+package seeder
+
+type Facade interface {
```
Comment (hwbrzzl) [inline contracts/database/seeder/seeder.go:3]: S: Add `Call` method.

File: database/seeder.go (hunk 12 above)
Comment (hwbrzzl) [reply database/seeder.go:13]: The `Called` can be moved to `SeederFacade `.

File: database/seeder.go [hunk 26]
```diff
@@ -0,0 +1,109 @@
...
+//
+//	seeder := &Seeder{}
+//	seeder.Call([]seeder.Seeder{&UserSeeder{}})
+//	seeder.Call([]seeder.Seeder{&UserSeeder{}, &PostSeeder{}})
+//
+// Parameters:
+//   - seeders ([]seeder.Seeder): The seeder class or a slice of seeder classes to execute.
+//
+// Returns:
+//   - error: An error if the execution fails.
```
Comment (hwbrzzl) [inline database/seeder.go:54]: S: Remove this.

File: database/seeder.go [hunk 27]
```diff
@@ -0,0 +1,109 @@
...
+	for _, seeder := range seeders {
+		name := fmt.Sprintf("%T", seeder)
+
+		if contains(s.Called, name) {
+			continue
+		}
+
+		err := seeder.Run()
+		if err != nil {
+			log.Println("Error executing seeder:", err)
```
Comment (hwbrzzl) [inline database/seeder.go:65]:
```suggestion
```

File: database/seeder.go [hunk 28]
```diff
@@ -0,0 +1,109 @@
...
+//   - error: An error if the execution fails.
+func (s *Seeder) CallOnce(seeders []seeder.Seeder) error {
+	seederType := reflect.TypeOf(seeders)
+	seederTypeName := seederType.String()
+	seederPointerTypeName := "*" + seederTypeName
+
+	for _, called := range s.Called {
+		if called == seederTypeName || called == seederPointerTypeName {
+			return nil
+		}
```
Comment (hwbrzzl) [inline database/seeder.go:105]: S: It seems that the logic is wrong, please add some unit tests to test it.

File: database/seeder.go
Comment (hwbrzzl) [inline database/seeder.go:1]: S: Please add unit tests for this file.

**Review by hwbrzzl — COMMENTED** [review]
We have a great improvement. 👍

File: contracts/database/seeder/seeder.go (hunk 20 above)
Comment (kkumar-gcc) [reply contracts/database/seeder/seeder.go:11]: done

File: database/console/stubs.go (hunk 22 above)
Comment (kkumar-gcc) [reply database/console/stubs.go:75]: done

File: database/console/stubs.go [hunk 29]
```diff
@@ -71,3 +71,54 @@ func (u *DummyObserver) ForceDeleted(event orm.Event) error {
...
+    //     go run . artisan db:seed --seeder DummySeeder
+	//
+	// The Run method performs the actual seeding operations.
+	// To register other seeders to run, use the CallOnce method and provide the seeder instances.
+	// Example:
+	//     // Register multiple seeders
+	//     s.CallOnce([]seeder.Seeder{&seeders.&OtherSeeder{},&seeders.&OtherSeeder2{}})
+	//
+	// Make sure to adjust the import statements and package paths based on your project structure.
+
```
[the remark this answers is not in the document]
Comment (kkumar-gcc) [reply database/console/stubs.go:120]: done

File: database/console/seed_command.go [hunk 30]
```diff
@@ -0,0 +1,109 @@
+package console
+
+import (
+	"errors"
+	"log"
+
+	_ "github.com/go-sql-driver/mysql"
+	_ "github.com/golang-migrate/migrate/v4/source/file"
```
[the remark this answers is not in the document]
Comment (kkumar-gcc) [reply database/console/seed_command.go:8]: removed

File: database/console/seed_command.go [hunk 31]
```diff
@@ -0,0 +1,109 @@
...
+		},
+	}
+}
+
+// Handle executes the console command.
+func (receiver *SeedCommand) Handle(ctx console.Context) error {
+	err := receiver.ConfirmToProceed(ctx)
+	if err != nil {
+		color.Redln(err)
+		return err
```
[the remark this answers is not in the document]
Comment (kkumar-gcc) [reply database/console/seed_command.go:64]: resolved

File: database/console/seed_command.go [hunk 32]
```diff
@@ -0,0 +1,109 @@
...
+
+	color.Greenln("Seeding database.")
+	names := ctx.OptionSlice("seeder")
+	seeders := receiver.GetSeeders(names)
+
+	for _, seeder := range seeders {
+		err := seeder.Run()
+		if err != nil {
+			log.Printf("Error running seeder: %v\n", err)
+			continue
```
[the remark this answers is not in the document]
Comment (kkumar-gcc) [reply database/console/seed_command.go:75]: I think we still can run other seeder if there is an error in a seeder.

File: database/console/seed_command.go (hunk 19 above)
Comment (kkumar-gcc) [reply database/console/seed_command.go:93]: done

File: database/console/seed_command.go [hunk 33]
```diff
@@ -0,0 +1,109 @@
...
+		log.Println("No seeders specified, running all seeders.")
+		return seeders.GetAllSeeder()
+	}
+	var seederInstances []seeder.Seeder
+	for _, name := range names {
+		class := "seeders." + name
+		seeder := seeders.GetSeeder(class)
+		if seeder == nil {
+			log.Printf("No seeder of type %s found\n", class)
+			continue
```
[the remark this answers is not in the document]
Comment (kkumar-gcc) [reply database/console/seed_command.go:100]: If we want to stop user to execute other seeder than it is better.

File: database/seeder.go (hunk 24 above)
Comment (kkumar-gcc) [reply database/seeder.go]: done

**Comment by hwbrzzl** [comment]
One more thing, please test this feature by goravel/goravel in your local.

**Comment by kkumar-gcc** [comment]
> One more thing, please test this feature by goravel/goravel in your local.

Yeah, I am testing.

File: database/console/seed_command.go [hunk 34]
```diff
@@ -0,0 +1,109 @@
...
+// GetSeeder returns a seeder instance from the container.
+func (receiver *SeedCommand) GetSeeders(names []string) []seeder.Seeder {
+	seeders := receiver.seeder
+	if len(names) == 0 {
+		log.Println("No seeders specified, running all seeders.")
+		return seeders.GetAllSeeder()
+	}
+	var seederInstances []seeder.Seeder
+	for _, name := range names {
+		class := "seeders." + name
```
[the remark this answers is not in the document]
Comment (kkumar-gcc) [reply database/console/seed_command.go]: removed seeders.

File: database/seeder.go (hunk 28 above)
Comment (kkumar-gcc) [reply database/seeder.go]: done

File: database/console/seed_command.go [hunk 35]
```diff
@@ -0,0 +1,96 @@
...
+// ConfirmToProceed determines if the command should proceed based on user confirmation.
+func (receiver *SeedCommand) ConfirmToProceed(ctx console.Context) error {
+	force := ctx.OptionBool("force")
+	if force || (receiver.config.Env("APP_ENV") != "production") {
+		return nil
+	}
+	return errors.New("application in production use --force to run this command")
+}
+
+// GetSeeder returns a seeder instance from the container.
```
Comment (hwbrzzl) [inline database/console/seed_command.go:81]:
```suggestion
// GetSeeders returns seeder instances
```

File: database/console/seed_command.go [hunk 36]
```diff
@@ -0,0 +1,96 @@
...
+func (receiver *SeedCommand) GetSeeders(names []string) []seeder.Seeder {
+	if len(names) == 0 {
+		return receiver.seeder.GetSeeders()
+	}
+	var seeders []seeder.Seeder
+	for _, name := range names {
+		seeder := receiver.seeder.GetSeeder(name)
+		if seeder == nil {
+			color.Redf("No seeder of type %s found\n", name)
+			return nil
```
Comment (hwbrzzl) [inline database/console/seed_command.go:91]: S: it's better to return an error in an internal function, to print the error in main function.
```suggestion
			return nil, fmt.Errorf("No seeder of %s found", name)
```

File: database/console/stubs.go (hunk 23 above)
Comment (hwbrzzl) [reply database/console/stubs.go:83]: S: I think we need to remove this line.

File: database/seeder.go [hunk 37]
```diff
@@ -0,0 +1,81 @@
...
+	Seeders []seeder.Seeder
+	Called  []string
+}
+
+func NewSeederFacade() seeder.Facade {
+	return &SeederFacade{}
+}
+
+func (s *SeederFacade) Register(seeders []seeder.Seeder) {
+	s.Seeders = append(s.Seeders, seeders...)
```
Comment (hwbrzzl) [inline database/seeder.go:20]: S: To check if there are repeat signatures, use `color.Redln` to print them.

File: database/seeder.go [hunk 38]
```diff
@@ -0,0 +1,81 @@
...
+
+		if !contains(s.Called, signature) {
+			s.Called = append(s.Called, signature)
+		}
+	}
+	return nil
+}
+
+// contains checks if a string exists in a slice.
+func contains(slice []string, str string) bool {
```
Comment (hwbrzzl) [inline database/seeder.go:57]: S: Move this function to bottom.

File: database/seeder.go [hunk 39]
```diff
@@ -0,0 +1,81 @@
...
+		signature := item.Signature()
+
+		if contains(s.Called, signature) {
+			return nil
+		}
+
+		err := s.Call([]seeder.Seeder{item})
+		if err != nil {
+			return err
+		}
```
Comment (hwbrzzl) [inline database/seeder.go:78]:
```suggestion
		if err := s.Call([]seeder.Seeder{item}); err != nil {
			return err
		}
```

**Review by hwbrzzl — COMMENTED** [review]
A great improvement, two more thing:
1. Update `testing/mock/mock.go` file;
2. Please add some screenshots of the test results in goravel/goravel;

File: database/console/migrate_refresh_command.go [hunk 40]
```diff
@@ -91,6 +102,15 @@ func (receiver *MigrateRefreshCommand) Handle(ctx console.Context) error {
...
 
+	// Seed the database if the "seed" flag is provided
+	if ctx.OptionBool("seed") {
+		seeders := ctx.OptionSlice("seeder")
+		seederFlag := ""
+		if len(seeders) > 0 {
+			seederFlag = "--seeder " + strings.Join(seeders, ",")
+		}
+		receiver.artisan.Call("db:seed " + seederFlag)
+	}
```
Comment (hwbrzzl) [inline database/console/migrate_refresh_command.go:113]: Ditto

File: database/console/seed_command.go [hunk 41]
```diff
@@ -0,0 +1,106 @@
+package console
+
+import (
+	"errors"
+	"fmt"
+
+	color "github.com/gookit/color"
```
Comment (hwbrzzl) [inline database/console/seed_command.go:7]:
```suggestion
	"github.com/gookit/color"
```

File: database/seeder_test.go [hunk 42]
```diff
@@ -0,0 +1,116 @@
...
+	"log"
+	"testing"
+
+	"github.com/goravel/framework/contracts/database/seeder"
+	"github.com/stretchr/testify/suite"
+)
+
+type SeederTestSuite struct {
+	suite.Suite
+	seederFacade seeder.Facade
```
Comment (hwbrzzl) [inline database/seeder_test.go:13]:
```suggestion
	seederFacade *SeederFacade
```

**Review by hwbrzzl — COMMENTED** [review]
We almost finish this PR, just a few tips. Great work, Bro!

File: database/seeder.go [hunk 43]
```diff
@@ -0,0 +1,88 @@
...
+	"github.com/gookit/color"
+	"github.com/goravel/framework/contracts/database/seeder"
+)
+
+type SeederFacade struct {
+	Seeders []seeder.Seeder
+	Called  []string
+}
+
+func NewSeederFacade() seeder.Facade {
```
Comment (hwbrzzl) [inline database/seeder.go:13]:
```suggestion
func NewSeederFacade() *SeederFacade {
```

File: database/console/migrate_fresh_command_test.go [hunk 44]
```diff
@@ -86,11 +86,23 @@ func TestMigrateFreshCommand(t *testing.T) {
...
-
+			mockArtisan := &consolemocks.Artisan{}
 			migrateCommand := NewMigrateCommand(mockConfig)
 			assert.Nil(t, migrateCommand.Handle(mockContext))
+			mockContext.On("OptionBool", "seed").Return(false).Once()
+			migrateFreshCommand := NewMigrateFreshCommand(mockConfig, mockArtisan)
+			assert.Nil(t, migrateFreshCommand.Handle(mockContext))
+
+			// Test MigrateFreshCommand with --seed flag and seeders specified
+			mockContext.On("OptionBool", "seed").Return(true).Once()
```
Comment (hwbrzzl) [inline database/console/migrate_fresh_command_test.go:97]: S: We'd better to create new mockContext, mockArtisan, and migrateFreshCommand

File: database/console/migrate_fresh_command_test.go [hunk 45]
```diff
@@ -86,11 +86,23 @@ func TestMigrateFreshCommand(t *testing.T) {
...
+			mockContext.On("OptionSlice", "seeder").Return([]string{"MockSeeder"}).Once()
+			mockArtisan.On("Call", "db:seed --seeder MockSeeder").Return(nil).Once()
+			assert.Nil(t, migrateFreshCommand.Handle(mockContext))
 
-			migrateFreshCommand := NewMigrateFreshCommand(mockConfig)
+			// Test MigrateFreshCommand with --seed flag and no seeders specified
+			mockContext.On("OptionBool", "seed").Return(true).Once()
```
Comment (hwbrzzl) [inline database/console/migrate_fresh_command_test.go:103]: Ditto

File: database/console/migrate_refresh_command_test.go [hunk 46]
```diff
@@ -102,8 +106,18 @@ func TestMigrateRefreshCommand(t *testing.T) {
 
 			migrateCommand = NewMigrateCommand(mockConfig)
 			assert.Nil(t, migrateCommand.Handle(mockContext))
+			// Test MigrateRefreshCommand with --seed flag and --seeder specified
```
Comment (hwbrzzl) [inline database/console/migrate_refresh_command_test.go:109]:
```suggestion
			
			// Test MigrateRefreshCommand with --seed flag and --seeder specified
```

File: database/console/migrate_refresh_command_test.go [hunk 47]
```diff
@@ -102,8 +106,18 @@ func TestMigrateRefreshCommand(t *testing.T) {
...
+			mockContext.On("OptionBool", "seed").Return(true).Once()
```
Comment (hwbrzzl) [inline database/console/migrate_refresh_command_test.go:110]: Ditto

File: database/console/migrate_refresh_command_test.go [hunk 48]
```diff
@@ -102,8 +106,18 @@ func TestMigrateRefreshCommand(t *testing.T) {
...
+			mockContext.On("OptionSlice", "seeder").Return([]string{"UserSeeder"}).Once()
+			mockArtisan.On("Call", "db:seed --seeder UserSeeder").Return(nil).Once()
+			assert.Nil(t, migrateRefreshCommand.Handle(mockContext))
+
+			mockContext.On("Option", "step").Return("").Once()
 
-			migrateRefreshCommand = NewMigrateRefreshCommand(mockConfig)
+			// Test MigrateRefreshCommand with --seed flag and no --seeder specified
+			mockContext.On("OptionBool", "seed").Return(true).Once()
```
Comment (hwbrzzl) [inline database/console/migrate_refresh_command_test.go:118]: Ditto

**Review by hwbrzzl — COMMENTED** [review]
I think we can go ahead after this review. 👍

**Review by hwbrzzl — APPROVED** [review]
Perfect, great work!

**Pull request merged**